### PR TITLE
Symbol: operator+; operator-;

### PR DESCRIPTION
**Pull request by jane-intel**
### Details:
 - *Introduces operator+ and operator- for ov::Symbol*
 - *Uses introduced operators in `Add::evaluate_symbol`, `Subtract::evaluate_symbol` and shape_infer of Range operation*

Significantly reduces the number of unique symbols in models, allowing better optimization of ShapeOf subgraphs in dynamic models.

### Tickets:
 - *CVS-112292*

### CODE REVIEW

File: src/bindings/python/src/pyopenvino/graph/symbol.cpp [hunk 1]
```diff
@@ -25,6 +25,20 @@ void regclass_graph_Symbol(py::module m) {
...
+        },
+        py::is_operator());
+
+    symbol.def(
+        "__sub__",
+        [](const std::shared_ptr<ov::Symbol>& a, const std::shared_ptr<ov::Symbol>& b) {
+            return a - b;
+        },
+        py::is_operator());
+
```
Comment (akuporos) [inline src/bindings/python/src/pyopenvino/graph/symbol.cpp:41]: Please also add/extend tests [here](https://github.com/openvinotoolkit/openvino/blob/master/src/bindings/python/tests/test_runtime/test_dimension.py#L96-L97)

File: src/core/include/openvino/core/symbol.hpp [hunk 2]
```diff
@@ -3,37 +3,93 @@
...
+    friend SymbolPtr ov::operator-(const SymbolPtr& lhs, const SymbolPtr& rhs);
 };
 
 }  // namespace ov
+
+namespace std {
+bool operator==(const ov::symbol::WeakSymbol& lhs, const ov::symbol::WeakSymbol& rhs);
+bool operator!=(const ov::symbol::WeakSymbol& lhs, const ov::symbol::WeakSymbol& rhs);
+bool operator<(const ov::symbol::WeakSymbol& lhs, const ov::symbol::WeakSymbol& rhs);
+}  // namespace std
```
Comment (jane-intel) [inline src/core/include/openvino/core/symbol.hpp:95]: to be removed from std namespace

**Review by akuporos — APPROVED** [review]

File: src/core/tests/symbol.cpp [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+//
+
+#include "openvino/core/symbol.hpp"
+
+#include <gtest/gtest.h>
+
+#include <memory>
+
+using namespace std;
+using namespace ov;
```
Comment (praasz) [inline src/core/tests/symbol.cpp:12]:
```suggestion
namespace ov {
namespace test {
```
Consider wrap test in this namespace, then using `ov::` would not be required and I think `using std` is also not required.

File: src/core/src/op/subtract.cpp [hunk 4]
```diff
@@ -63,6 +63,21 @@ bool Subtract::evaluate(TensorVector& outputs, const TensorVector& inputs) const
                                       get_autob());
 }
 
+bool Subtract::evaluate_symbol(ov::TensorSymbolVector& output_symbols) const {
+    const auto &lhs_pshape = input(0).get_tensor().get_partial_shape(),
+               rhs_pshape = input(1).get_tensor().get_partial_shape();
```
Comment (praasz) [inline src/core/src/op/subtract.cpp:68]:
```suggestion
               &rhs_pshape = input(1).get_tensor().get_partial_shape();
```
?
currently the `rhs_pshape` is not a reference.

**Review by praasz — APPROVED** [review]

**Review by rkazants — CHANGES_REQUESTED** [review]
TF layer test fails. Need to check

**Comment by rkazants** [comment]
Re-run does not help. The same test is failing:
FAILED install\tests\layer_tests\tensorflow_tests\test_tf_HSVToRGB.py::TestHSVToRGB::test_hsv_to_rgb_basic[ ie_device:CPU - precision:FP16 - params:{'input_shape': [7, 7, 3], 'input_type': <class 'numpy.float32'>, 'special_case': 'Black Image'} ]
FAILED install\tests\layer_tests\tensorflow_tests\test_tf_HSVToRGB.py::TestHSVToRGB::test_hsv_to_rgb_basic[ ie_device:CPU - precision:FP16 - params:{'input_shape': [7, 7, 3], 'input_type': <class 'numpy.float32'>, 'special_case': 'Grayscale Image'} ]
FAILED install\tests\layer_tests\tensorflow_tests\test_tf_HSVToRGB.py::TestHSVToRGB::test_hsv_to_rgb_basic[ ie_device:CPU - precision:FP16 - params:{'input_shape': [3, 4, 13, 15, 3], 'input_type': <class 'numpy.float64'>} ]
FAILED install\tests\layer_tests\tensorflow_tests\test_tf_HSVToRGB.py::TestHSVToRGB::test_hsv_to_rgb_basic[ ie_device:CPU - precision:FP16 - params:{'input_shape': [5, 23, 27, 3], 'input_type': <class 'numpy.float64'>} ]
FAILED install\tests\layer_tests\tensorflow_tests\test_tf_HSVToRGB.py::TestHSVToRGB::test_hsv_to_rgb_basic[ ie_device:CPU - precision:FP16 - params:{'input_shape': [5, 5, 3], 'input_type': <class 'numpy.float32'>} ]
= 5 failed, 2454 passed, 35 skipped, 19 xfailed, 1 xpassed, 40 warnings in 464.21s (0:07:44) =

Also, PR can be guilty because decomposition for HSVToRGB uses Add, Subtract operations.

**Comment by itikhono** [comment]
the issues with ONNX models are sporadic, passed after several re-runs and rebase.
the issues with TF models (test_tf_HSVToRGB) are not reproduced on ubuntu,  I'm trying to reproduce them on Windows

**Comment by rkazants** [comment]
> the issues with ONNX models are sporadic, the tests passed after several re-runs and rebase. the issues with TF models (test_tf_HSVToRGB) are not reproduced on ubuntu, I'm trying to reproduce them on Windows

@itikhono, @mlukasze 
Please don't re-run. It does not help. Failure on Windows is permanently reproduced. It needs to be fixed in this PR.

**Comment by mlukasze** [comment]
roger, roger

**Comment by itikhono** [comment]
the issue was reproduced on Windows, it looks like it's related to the hash calculation, investigating.

File: src/core/src/symbol.cpp [hunk 5]
```diff
@@ -4,27 +4,431 @@
...
+        size_t seed = 0;
+        for (const auto& element : v) {
+            const auto& el_hash = element.expired() ? 0 : std::hash<ov::SymbolPtr>()(element.lock());
+            seed ^= el_hash + 0x9e3779b9 + (seed << 6) + (seed >> 2);
+        }
+        return seed;
+    }
+};
+
+using MathMap = std::unordered_map<WeakSymbolVector, WeakSymbol, WeakSymbolVectorHash>;
```
Comment (itikhono) [inline src/core/src/symbol.cpp:41]:
```suggestion
using MathMap = std::map<WeakSymbolVector, WeakSymbol>;
```

File: src/core/src/symbol.cpp [hunk 6]
```diff
@@ -4,27 +4,431 @@
 
 #include "openvino/core/symbol.hpp"
 
-std::shared_ptr<ov::Symbol> ov::symbol::ancestor_of(const std::shared_ptr<Symbol>& symbol) {
+#include <algorithm>
+#include <queue>
```
Comment (itikhono) [inline src/core/src/symbol.cpp:8]:
```suggestion
#include <queue>
#include <map>
```

File: src/core/src/symbol.cpp [hunk 7]
```diff
@@ -4,27 +4,432 @@
...
+#include <map>
```
Comment (itikhono) [inline src/core/src/symbol.cpp:9]:
```suggestion
#include <map>
#include <queue>
```

**Comment by itikhono** [comment]
the issue is caused by the hash calculation algo:
image
this calculations might give us different results depends on the weak_ptrs are alive or not. 
If we try to use unordered_map with a key which might have different hash, it might cause seg fault.
So I replaced unordered_map with map. It doesn't rely on the hash function, and uses the overload < operator. 
We will be able to merge this PR after that.
 
But why weak_ptrs are expired is an another question. 
Most probably this is something similar to the issue fixed in this commit https://github.com/openvinotoolkit/openvino/pull/25360/commits/ef94d8cd6cb10ff8a4c9ecaf4d8c3ad686a9f556

**Review by rkazants — APPROVED** [review]

**Pull request merged**